Revise tech docs in README to favour GOV.UK Docker (#845)
alphagov/contacts-admin: README.md

## Technical documentation

> This is a Ruby on Rails app, and should follow [our Rails app conventions](https://docs.publishing.service.gov.uk/manual/conventions-for-rails-applications.html).

Reviewer: Commit 36dda45fc8d13cbf2256eb6bc80c91856d00eefa - the commit message references link `[1]` but doesn't provide a link 🤔

Author: Oops, thanks for spotting. Fixed.

The app has its own database (MySQL), which is used by the admin UI. This is in contrast to the separate [government-frontend](https://github.com/alphagov/government-frontend) app which reads contacts from the [content-store](https://github.com/alphagov/content-store). The admin UI part of this app handles publishing the contacts to the [publishing-api](https://github.com/alphagov/publishing-api) so that they are present for [finder-frontend](https://github.com/alphagov/finder-frontend) and [government-frontend](https://github.com/alphagov/government-frontend) to read.
You can use the [GOV.UK Docker environment](https://github.com/alphagov/govuk-docker) to run the application and its tests with all the necessary dependencies. Follow [the usage instructions](https://github.com/alphagov/govuk-docker#usage) to get started.

### Dependencies
**Use GOV.UK Docker to run any commands that follow.**

* Ruby 2.3.0
* MySQL
* [whitehall](https://github.com/alphagov/whitehall)
  1. to access the organisations API and maintain parity with the organisations managed in whitehall.

>   2. to access the world locations API to provide the correct list of countries for the country dropdown in the address form

Reviewer: Whilst I agree we can remove the `Dependencies` section, it might be nice to leave some of this useful context:

```suggestion
Contacts Admin uses Whitehall to access the organisations API (maintaining parity with the organisations managed in Whitehall) and to access the world locations API (to provide the correct list of countries for the country dropdown in the address form).
```

Without it, the developer has to look in the code to discover the (surprising) relationship with Whitehall. I think integrations with Signon and Content Store are pretty standard, but the Whitehall dependency is a snowflake that it can't hurt to document, IMO.

Author: Fair enough. I've amended the first commit with tweaked wording. I think this goes better in the "before running the app" section, as it's only relevant there.

> ### Before running ths app

Reviewer: ```suggestion
### Before running the app
```

> The best way to get a database with good seed data is to replicate from Integration. Alternatibely, you can use a combination of import commands and old seed data to seed the database.

Reviewer: ```suggestion
The best way to get a database with good seed data is to replicate from Integration. Alternatively, you can use a combination of import commands and old seed data to seed the database, as follows:
```

> env GOVUK_APP_DOMAIN='https://www.gov.uk' env GOVUK_WEBSITE_ROOT='https://www.gov.uk' bin/rake db:seed

Reviewer: I know the idea is to avoid using `govuk-docker-run` throughout the README, but in this case I think it's worth keeping the prefix in, as `env` is a Docker CLI argument so makes no sense in isolation.

Author: So [`env`](https://en.wikipedia.org/wiki/Env) is actually a normal UNIX command. We have to use it with Docker since other alternatives don't work:

```
VAR=value govuk-docker-run ... # env var not passed through

govuk-docker-run VAR=value ... # error, not a command
```

`env` is nice because it works with and without GOV.UK Docker, so I think it's reasonable to keep here.

Reviewer: Oops - I really should have Googled that first 😄  I haven't seen `env` used outside of Docker commands. Thanks for the explanation!

  command-line based tasks to withdraw and redirect a contact
### Before running ths app

### Running with govuk-docker

Setup:

```sh
govuk-docker-run env RAILS_ENV=test env bin/rake db:create
```

Then run tests:
The best way to get a database with good seed data is to replicate from Integration. Alternatibely, you can use a combination of import commands and old seed data to seed the database.

```sh
govuk-docker-run env RUNNING_IN_CI='true' bundle exec rspec
```

#### Previewing the application

Seed the database, with ENV overrides so that `ImportOrganisations` doesn't attempt to fetch organisations from dev:

```sh
govuk-docker-run env GOVUK_APP_DOMAIN='https://www.gov.uk' env GOVUK_WEBSITE_ROOT='https://www.gov.uk' bin/rake db:seed
```

Run the app:

```sh
govuk-docker-up
env GOVUK_APP_DOMAIN='https://www.gov.uk' env GOVUK_WEBSITE_ROOT='https://www.gov.uk' bin/rake db:seed
bundle exec rake contacts:import_hmrc DATA_FILE=db/contact-records.csv
```

Visit http://contacts-admin.dev.gov.uk/.

You may see sometimes see a `GdsApi::TimedOutException` when performing a CRUD operation such as creating a contact.

> This usually means Publishing API is still starting up. Wait a few moments, then try again.

Reviewer: Slightly biased as I wrote this line 😅 Agreed it is better centrally documented. I've found similar while running a local rake task in Frontend. I'll raise a PR to describe this in govuk-docker.

Author: Sounds good, thanks!

Reviewer: https://github.com/alphagov/govuk-docker/pull/462 👀 🙏

## Technical documentation

This is a Ruby on Rails app, and should follow [our Rails app conventions](https://docs.publishing.service.gov.uk/manual/conventions-for-rails-applications.html).

You can use the [GOV.UK Docker environment](https://github.com/alphagov/govuk-docker) to run the application and its tests with all the necessary dependencies. Follow [the usage instructions](https://github.com/alphagov/govuk-docker#usage) to get started.

**Use GOV.UK Docker to run any commands that follow.**

### Before running the app

The best way to get a database with good seed data is to replicate from Integration. Alternatively, you can use a combination of import commands and old seed data to seed the database, as follows:

```sh

> env GOVUK_APP_DOMAIN='https://www.gov.uk' env GOVUK_WEBSITE_ROOT='https://www.gov.uk' bin/rake db:seed

Reviewer 2: ```suggestion
env GOVUK_APP_DOMAIN='https://www.gov.uk' GOVUK_WEBSITE_ROOT='https://www.gov.uk' bin/rake db:seed
```

`env` accepts multiple environment variables, so it shouldn't be necessary to run `env` again.

Author: Indeed, nice catch. I've snuck this in.

bundle exec rspec
```

## Redirecting removed contacts
### Further documentation

> The Web UI of the app issues a “gone” item to the publishing api when a contact is deleted. Sometimes the correct thing to do is to issue a “redirect” instead. There is no web UI for this but there are two rake tasks to achieve this depending on the state of the contact to be redirected.

Reviewer: Oh, this is completely different as of #838 - contacts are ALL redirected, rather than `gone`. We can probably just delete this section!

...and, come to think of it, some of the rake tasks 🤔 

I'll raise a PR to do that - should have done it in #838 - and then we can rebase this PR.

Reviewer: PR: #846. 👀 🙏
